### DISCUSSION TITLE
Managing test support code (mocks, spies, etc) at scale

### DISCUSSION THREAD

**Discussion by andre-alves** [MobileNativeFoundation/discussions#202] [General]
Hello 👋,

I'm curious to know how folks with huge iOS code bases and many test targets organize their setup to handle test support code like mocks and spies. I think this [library](https://github.com/pointfreeco/xctest-dynamic-overlay) explains the challenges so let me quote it:

<blockquote>
It is very common to write test support code for libraries and applications. This often comes in the form of little domain-specific functions or helpers that make it easier for users of your code to formulate assertions on behavior.

Currently there are only two options for writing test support code:

- Put it in a test target, but then you can't access it from multiple other test targets. For whatever reason test targets cannot be imported, and so the test support code will only be available in that one single test target.
- Create a dedicated test support module that ships just the test-specific code. Then you can import this module into as many test targets as you want, while never letting the module interact with your regular, production code.

Neither of these options is ideal. In the first case you cannot share your test support, and the second case will lead you to a proliferation of modules. For each feature you potentially need 3 modules: MyFeature, MyFeatureTests and MyFeatureTestSupport. SPM makes managing this quite easy, but it's still a burden.

It would be far better if we could ship the test support code right along side or actual library or application code. After all, they are intimately related. You can even fence off the test support code in #if DEBUG ... #endif if you are worried about leaking test code into production.
</blockquote>

We are currently following the `Create a dedicated test support module that ships just the test-specific code` approach, but considering if we should just ship it with the actual library as the author suggests. Having hundreds of these test support modules do increase the maintenance burden of our test targets.

How do you handle these challenges?

**Comment by rockbruno** [comment]
Hey André :)

Folks who have built this integration may correct me on the specifics. At Spotify we use the test support method you mentioned. Yes, this initially leads to a lot of module bloat, but with:

1) Sufficient project generation support to hide the overhead of creating and linking additional modules, and
2) Remote build caching to combat the impact of all those modules on build times, especially on the CI

I personally think it brings those potential problems to manageable levels. It might also be cool to mention that we even have linters that try to detect the creation of public types and fails if it can't find an equivalent fake in the test support target.

Personally I'm skeptical of the "ship everything together under macros" approach, that sounds very hard to manage at scale.

**Reply by andre-alves** [reply]
Nice to hear that folks are doing something similar!

One thing that might change the way we manage test support code is Swift Macros. Some [libs](https://github.com/Matejkob/swift-spyable) are already being created to automate spies. If this turns out to be viable, then the future might be shipping everything together.
